src/array-repeat-strategy.js

```diff
@@ -20,16 +20,82 @@ export class ArrayRepeatStrategy {
   * @param items The new array instance.
   */
   instanceChanged(repeat, items) {
-    if (repeat.viewsRequireLifecycle) {
-      let removePromise = repeat.removeAllViews(true);
-      if (removePromise instanceof Promise) {
-        removePromise.then(() => this._standardProcessInstanceChanged(repeat, items));
+    const itemNameInBindingContext = repeat.local;
+    
+    const children = repeat.viewSlot.children;
+    let itemsPreviouslyInViews = children.map(view => view.bindingContext[itemNameInBindingContext]);
+    
+    let itemsLength = items.length;
+    
+    // remove items no longer in the instance
+    if (itemsPreviouslyInViews.length > 0) {
+      let indiciesToRemove = [];
+      
+      itemsPreviouslyInViews.forEach((oldItem, index) => {
+        if (items.indexOf(oldItem) === -1) {
+          indiciesToRemove.push(index);
+        }
+      })
+      let rmPromises = [];
+      
+      for (let i = indiciesToRemove.length - 1; i >= 0; i--) {
+        let viewOrPromise = repeat.removeView(indiciesToRemove[i], true);
+        if (viewOrPromise instanceof Promise) {
+          rmPromises.push(viewOrPromise);
+        }
+        
+        // remove also from the cache list:
+        itemsPreviouslyInViews.splice(indiciesToRemove[i], 1);
+      }
+      
+      // update views
+      Promise.all(rmPromises).then(() => {
+        let index = 0;
+        for (let item of items) {
```

Comment:
We want to avoid `for..of` loops throughout the Aurelia codebase since the compiled code uses Symbols. Please change to a good old for-loop.